src/js/angular/guides/tour-lib-services/shepherd.service.js

```diff
@@ -0,0 +1,543 @@
+                this._clearLocaleStorage();
+                if (this.onCancel) {
+                    this.onCancel();
+                }
+            });
+        }
+    }
+
+    /**
+     * Creates first step from <code>stepsDescriptions</code>.
+     * @param guide - the guide.
+     * @param stepsDescriptions - array with descriptions of steps.
+     * @returns the first step.
+     * @private
+     */
+    this._getFirstStep = function (guide, stepsDescriptions) {
+        const nextStepDescription = stepsDescriptions.length > 1 ? stepsDescriptions[1] : undefined;
+        return this._toGuideStep(guide, undefined, stepsDescriptions[0], nextStepDescription);
+    }
+
+    /**
+     * Creates middle step from <code>stepsDescriptions</code>.
+     * @param guide - the guide.
+     * @param stepsDescriptions - array with descriptions of steps.
+     * @param indexOfProcessedStep - index of step which have to be created.
+     * @returns the created step.
+     * @private
+     */
+    this._getMiddleStep = (guide, stepsDescriptions, indexOfProcessedStep) => {
+        return this._toGuideStep(guide, stepsDescriptions[indexOfProcessedStep - 1], stepsDescriptions[indexOfProcessedStep], stepsDescriptions[indexOfProcessedStep + 1]);
+    }
+
+    /**
+     * Creates last step from <code>stepsDescriptions</code>.
+     * @param guide - the guide.
+     * @param stepsDescriptions - array with descriptions of steps.
+     * @returns the last step.
+     * @private
+     */
+    this._getLastStep = (guide, stepsDescriptions) => {
+        return this._toGuideStep(guide, stepsDescriptions[stepsDescriptions.length - 2], stepsDescriptions[stepsDescriptions.length - 1]);
+    }
+
+    /**
+     * Converts <code>currentStepDescription</code> to Shepherd Tour step.
+     * @param guide - the guide.
+     * @param previousStepDescription - the previous step description.
+     * @param currentStepDescription - processed step description.
+     * @param nexStepDescription - the next step description.
+     * @returns created step.
+     * @private
+     */
+    this._toGuideStep = (guide, previousStepDescription, currentStepDescription, nexStepDescription) => {
+        const step = this._toBaseGuideStep($translate, currentStepDescription, () => this._updateLocalStorage());
+        const buttons = [];
+        buttons.push(this._getPreviousButton(guide, previousStepDescription, currentStepDescription));
+        buttons.push(this._getNextButton(guide, currentStepDescription, nexStepDescription));
+        buttons.push(this._getPauseButton(guide));
+        step.buttons = buttons;
+        return step;
+    }
+
+    this._getPauseButton = (guide) => {
+        return this._getButton($translate.instant('guide.button.pause'), () => this._pauseGuide(guide), true);
+    }
+
+    /**
+     * Pauses current ran guide.
+     */
+    this._pauseGuide = (guide) => {
+        guide.hide();
+        this._getStepWhichCanBePaused(Shepherd.activeTour);
```

Comment:
What do you do with the step which can be paused when you find it?

Reply:
We have to save the step. Fixed